File: javascript/20-sum-factorial-digits.js
function multiply(number, x) {
  let result = [];
  let carry = 0;

  for (let i = 0; i < number.length; i++) {
    let temp = number[i] * x + carry;
    carry = Math.floor(temp / 10);
    result.push(temp % 10);
  }

  while (carry > 0) {
    result.push(carry % 10);
    carry = Math.floor(carry / 10);
  }

  return result;
}

function sumFactorialDigits(n) {
  let number = [1];

  for (let i = 1; i <= n; i++) {
    number = multiply(number, i);
  }

  let result = 0;
  for (let i = 0; i < number.length; i++) {
    result += number[i];
  }

  return result;
}

console.log(sumFactorialDigits(100));

File: go/20-sum-factorial-digits.go
package main

import "fmt"

func multiply(number []int, x int) []int {
	result := []int{}
	carry := 0

	for i := 0; i < len(number); i++ {
		temp := number[i]*x + carry
		carry = temp / 10
		result = append(result, temp%10)
	}

	for carry > 0 {
		result = append(result, carry%10)
		carry = carry / 10
	}

	return result
}

func sumFactorialDigits(n int) int {
	number := []int{1}

	for i := 1; i <= n; i++ {
		number = multiply(number, i)
	}

	sum := 0
	for i := 0; i < len(number); i++ {
		sum += number[i]
	}

	return sum
}

func main() {
	fmt.Println(sumFactorialDigits(100))
}
